Бравилов Николай

kol9nus opened the pull request.

kol9nus committed:
Add functionality

honest-hrundel commented:
:tomato: Пройдено тестов 14 из 18

kol9nus committed:
Get all possible properties

There was properties of first person only.

honest-hrundel commented:
:tomato: Пройдено тестов 14 из 18

kol9nus committed:
Add some checks

honest-hrundel commented:
:tomato: Пройдено тестов 14 из 18

kol9nus committed:
Add limit reset

It limits all queries without limit by previous limit, if it's not
resetting

honest-hrundel commented:
:tomato: Пройдено тестов 15 из 18

kol9nus committed:
Add operations priorities

honest-hrundel commented:
:tomato: Пройдено тестов 17 из 18

kol9nus committed:
Add redefinition of limit

honest-hrundel commented:
:tomato: Пройдено тестов 17 из 18

kol9nus committed:
Change checker

honest-hrundel commented:
:tomato: Пройдено тестов 17 из 18

kol9nus committed:
Ignore unexisting properties in selects

honest-hrundel commented:
:tomato: Пройдено тестов 17 из 18

kol9nus committed:
Revert "Ignore unexisting properties in selects"

This reverts commit 73b009c3d01779da060246bad8f886a39968d602 'cause
of unnecessary functionality.

kol9nus committed:
Implement indexes instead of copy

honest-hrundel commented:
:tomato: Пройдено тестов 17 из 18

kol9nus committed:
Add filtered arrays immutability

honest-hrundel commented:
:tomato: Пройдено тестов 17 из 18

kol9nus committed:
Brutforce or

honest-hrundel commented:
:green_apple: Пройдено тестов 18 из 18

honest-hrundel commented:
@Mokoshka обрати внимание решено доп. задание

kol9nus committed:
Return to non-indexing filtering

kol9nus committed:
Implement bruteforce "or" filtering

honest-hrundel commented:
:green_apple: Пройдено тестов 18 из 18

honest-hrundel commented:
@Mokoshka обрати внимание решено доп. задание

Mokoshka commented on lego.js:
```diff
@@ -6,65 +6,180 @@
- * @param {Array} collection
- * @params {...Function} – Функции для запроса
+ * @param {Array} friends
+ * @params {...{priority: number, operator: function}}
```
А почему ты изменил приходящие аргументы?

kol9nus replied:
Потому что каждая функция (sortBy, filterIn....) возвращает объект состоящий из priority, являющийся числом и operator, функция, которая непосредственно что-то делает с результатом. А collection во friends переименовал, показалось, что так логичнее

Mokoshka replied:
Ну со friends ты погорячился - этот код не только для друзей может вызываться. Коллекции могут быть разными, но обработка будет одна и та же.
Про объект с priority и operator я еще посмотрю по коду, возможно можно будет поправить код, не меняя  аргументы.

Mokoshka commented on lego.js:
```diff
@@ -6,65 +6,180 @@
+function clone(collection1, collection2) {
+    collection1.forEach(function (element) {
+        collection2.push({});
+        Object.keys(element).forEach(function (key) {
```
Здесь можно reduce использовать

Mokoshka commented on lego.js:
```diff
@@ -6,65 +6,180 @@
+ * @param {Array} collection1 - array of objects
+ * @param {Array} collection2 - array of objects
+ */
+function clone(collection1, collection2) {
```
Хм, кажется, функция копирования должна принимать один объект и возвращать его копию. Это же не ассемблер) Учитывая это, подумай как можно улучшить код функции

kol9nus replied:
Ну да, логично:D

Mokoshka commented on lego.js:
```diff
@@ -6,65 +6,180 @@
+        priority: OPERATORS_PRIORITIES.sortBy,
+        operator: function (friends) {
+            return friends.sort(function (friend1, friend2) {
+                return order === 'asc' ? friend1[property] > friend2[property]
```
Что это за магическая строка?

kol9nus replied:
Дак тернарник же. Если требуют по возрастанию, то сортировать по возрастанию, если не требуют по возрастанию, значит требуют по убыванию, то сортировать по убыванию..

Mokoshka replied:
Я про строку 'asc', может вынести ее в константу? Сейчас совсем непонятно, что это за три буквы

Mokoshka commented on lego.js:
```diff
@@ -6,65 +6,180 @@
+
+    executionQueue.forEach(function (priorityFunctions) {
+        priorityFunctions.forEach(function (operator) {
+            selectedFriends = operator(selectedFriends);
```
Как-то здесь сложно. Тебе нужно создать очередь, потом каждый ее элемент-функцию вызывать, используя всего один forEach. Очередь можно отсортировать, чтобы задать нужный порядок. Но двумерный массив кажется лишним здесь

kol9nus replied:
Да, но это максимально эффективный код для моих целей.

kol9nus replied:
Массив не лишний, мне кажется, потому что, если мы сделаем очередь и будем в ней сортировать, то мы будем сравнивать между собой элементы с одинаковым приоритетом (ещё и предварительно переписав в неё все элементы, кроме случая, если сортировать при добавлении). В случае с двумерным массивом, мы сразу скидываем в нужный уровень. Т.е. для двумерного массива сложность будет O(n) (просто переписываем все функции сразу куда надо), а для очереди с сортировкой как минимум O(2n) (как минимум считать - один проход, отсортировать как минимум один проход (если уже отсортирован)). Может я конечно перемудрил..

Mokoshka commented on lego.js:
```diff
@@ -6,65 +6,180 @@
+            friends.forEach(function (friend) {
+                Object.keys(friend).forEach(function (friendProperty) {
+                    if (selectors.indexOf(friendProperty) === -1) {
+                        delete friend[friendProperty];
```
Так себе идея удалять объекты из коллекции. Может лучше возвращать новую коллекцию с нужными данными?

kol9nus replied:
Да, точно, они же по ссылке передаются.

Mokoshka commented on lego.js:
```diff
@@ -6,65 +6,180 @@
+        priority: OPERATORS_PRIORITIES.select,
+        operator: function (friends) {
+            friends.forEach(function (friend) {
+                Object.keys(friend).forEach(function (friendProperty) {
```
Вопрос на подумать: действительно ли надо пробегаться по ВСЕМ полям записи? Может можно чуть-чуть оптимизировать?

kol9nus replied:
Ну единственная оптимизация, которую я вижу, выбирать массив с наименьшим количеством элементов среди полей записи и полям селекта. Мы же не знаем насколько большой селект к нам придёт, чтобы пробегаться по нему. А поиск пересечения этих массивов всё равно пройдётся по всем полям записи.

Mokoshka replied:
Но так же нам неизвестно сколько полей в объекте. Что подсказывает здравый смысл, где вероятность встретить меньшее количество полей? Кажется, что в коллекции в большинстве случаев полей будет больше, возможно намного. Может я и не права

kol9nus replied:
Ну, если Билли адекватный, спору нет, но кто знает современных пользователей..

kol9nus replied:
Я тут подумал, а есть ли разница, по чему пробегаться? Мы же после этого всё равно проверяем наличие текущего элемента (в цикле forEach) во втором массиве (т.е. пробегаемся и по нему). Если конечно hasOwnProperty делает это как то эффективно, а не просматривает все ключи..

Mokoshka commented on lego.js:
```diff
@@ -6,65 +6,180 @@
+                return values.some(function (propertyValue) {
+                    return friend[property] === propertyValue;
+                });
+            });
```
Большая вложенность. Можно немного уменьшить

Mokoshka commented:
🍅

kol9nus committed:
Refactor code

Cause of 'select' function is really similar to 'clone' function
it is logical to do select through clone.

honest-hrundel commented:
:green_apple: Пройдено тестов 18 из 18

honest-hrundel commented:
@Mokoshka обрати внимание решено доп. задание

Mokoshka commented on lego.js:
```diff
@@ -6,65 +6,215 @@
-    return collection;
+    var queryResult = [];
+    uniqueCollectionKeys = getAllUniqueKeys(collection);
+    queryResult = clone(collection);
```
Обязательно ли queryResult сначала объявлять пустым массивом?

Mokoshka commented on lego.js:
```diff
@@ -6,65 +6,215 @@
+        priority: OPERATORS_PRIORITIES.format,
+        operator: function (collection) {
+            collection.forEach(function (item) {
+                if (item[property] !== undefined) {
```
Обязательно сравнивать с undefined? Может лучше переписать, используя метод объектов hasOwnProperty()?

Mokoshka commented on lego.js:
```diff
@@ -6,65 +6,215 @@
  * @param {Number} count – Максимальное количество элементов
+ * @returns {{priority: number, operator: function}}
  */
 exports.limit = function (count) {
```
А если передадут отрицательный count?

kol9nus replied:
Дак так же можно много где поломать всё.. Я делал в расчёте на корректные данные.

Mokoshka replied:
Отрицательные значения не такие уж и некорректные данные. Это не случай, где тебе например вместо объекта внезапно придет массив, например)
Давай, если здесь будет отрицательное число, то вернется 0 записей

Mokoshka commented on lego.js:
```diff
@@ -6,65 +6,215 @@
+    return {
+        priority: OPERATORS_PRIORITIES.select,
+        operator: function (collection) {
+            return clone(collection, getCorrectProperties(selectors));
```
Мне кажется или стало только сложнее? Как минимум стал менее читабельный и понятный код. А если в коллекции объекты с достаточно разными полями, то массив uniqueCollectionKeys раздует по размерам. + глобальная переменная не есть хорошо. Так же добавились еще итерации. Не стоит жертвовать читабельностью кода ради мелких оптимизаций.
Лучше оставить как было. Там была проблема лишь в том, чтобы понять, какой массив будет меньше. Как показывает практика, то в объектах содержится, как правило, много полей, а полей, по которым искать и выводить, достаточно мало. Очень редко встречаются случаи когда, в объекте полей меньше, чем селекторов

kol9nus replied:
Да, он стал немного запутаннее, но, по-моему логичнее. Со стороны, не зная контекста, может быть легко запутаться.
Просто есть принцип DRY, который я бы нарушил, т.к. clone по логике делает то же самое, что и select (в контексте предыдущего комментария, желательно не удалять поля, а создавать объект с нужными), но select делает это для определённых полей. Вот я и решил это имплементировать в clone (ну входной аргумент properties)

Mokoshka replied:
Во всем нужна мера же, простая функция clone стала совсем не простой. По мне, так clone должен возвращать копию объекта, а здесь нужно отфильтровать по ключам. А сейчас clone и копирует, и фильтрует

kol9nus replied:
Да, согласен, перемудрил.

Mokoshka commented on lego.js:
```diff
@@ -6,65 +6,215 @@
+function getAllUniqueValues(object) {
+    return Object.keys(object).reduce(function (uniqueValues, key, index) {
+        return uniqueValues.concat(
+            uniqueValues.indexOf(object[key]) !== index ? object[key] : []
```
Как-то сложно. Почему нельзя просто в массив добавлять значение, если его нет в этом массиве?

Mokoshka commented on lego.js:
```diff
@@ -6,65 +6,215 @@
+ * @returns {Array} - глубокая копия collection
+ */
+function clone(collection, properties) {
+    var getSharedProperties = properties !== undefined
```
Просто properties

Mokoshka commented:
🍅

kol9nus committed:
Refactor code

honest-hrundel commented:
:green_apple: Пройдено тестов 18 из 18

honest-hrundel commented:
@Mokoshka обрати внимание решено доп. задание

Mokoshka commented:
🚀
